Track card reader battery property card reader events

samiuelson opened the pull request:
Logging card reader's battery level tracking card reader events.

Closes: #7651

### Description
1. This PR adds the `CardReaderTrackingInfoKeeper::setCardReaderBatteryLevel` method for storing battery-level data in memory. 
2. Battery level is added as a property to events related to the card reader (tracked using `CardReaderTracker::track` method). The name of the event's property: `card_reader_battery_level`.
3. The battery level gets updated immediately after the reader is connected and then during the payment flow we subscribe to battery level changes and update it continuously using `CardReaderTrackingInfoKeeper`.

### Testing instructions
**1. Observe that events related to the card reader include the `card_reader_battery_level`.** 
Note: it won't be included if there are no battery-level data available. Usually, this shouldn't happen.
**2. Observe that the battery level is updated.** 
Eg. This can happen during long-running card payment flow if the reader is not connected to the cable. In that case, the battery level tracked with `*_card_reader_connection_success` event would differ from `*_card_present_collect_payment_success`. Note: the time from connecting to a reader to tapping a card must be long enough to observe battery level change.

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

samiuelson committed:
Log battery level when reader is connected

Update battery level when card reader gets connected and track it with connection successful event.

samiuelson committed:
Listen to reader's battery during payment

Subscribe to reader's battery level changes and update its value in `cardReaderTrackingInfoKeeper`.

samiuelson committed:
Verify battery level being set in `CardReaderConnectViewModel`

samiuelson committed:
Verify battery level being continuously updated

Test if during the payment flow on each battery status event we update it's value in `CardReaderTrackingInfoKeeper`.

samiuelson committed:
Remove unused imports

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTracker.kt:
```diff
@@ -91,6 +92,13 @@ class CardReaderTracker @Inject constructor(
         }
     }
 
+    private fun addCardReaderBatteryLevelProperty(properties: MutableMap<String, Any>) {
```
❓ We could write tests to verify that the card reader tracker adds battery level property when the battery status is not null and doesn't add battery level property when the battery status is null in `CardReaderTrackerTest`. WDYT?

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderPaymentViewModelTest.kt:
```diff
@@ -2943,6 +2946,25 @@ class CardReaderPaymentViewModelTest : BaseUnitTest() {
     //endregion - Interac Refund tests
 
+    @Test
+    fun `when new battery status event is received, then tracking is updated with new battery level`() =
```
❓ We could write another test that verifies that the `cardReaderTrackingInfoKeeper.setCardReaderBatteryLevel` won't be called if the `CardReaderBatteryStatus` is anything but `StatusChanged`. WDYT?

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackingInfoProvider.kt:
```diff
@@ -41,13 +42,18 @@ class CardReaderTrackingInfoImpl @Inject constructor() : CardReaderTrackingInfoP
     }
+
+    override fun setCardReaderBatteryLevel(batteryLevel: Float?) {
+        trackingInfoInternal = trackingInfoInternal.copy(cardReaderBatteryLevel = batteryLevel)
```
💡 Sorry, I forgot to mention this during our pair programming session. The battery level will be displayed in float and might not be readable when analyzing the events. For example: `"card_reader_battery_level":0.05`

We have an extension function that converts the battery level from the float into a readable string. It's in the `CardReaderDetailViewModel` class. The extension is named `getReadersBatteryLevel()`. We could probably use this.

We might need to extract the core battery conversion logic of float to int into a separate reusable method though since the above extension method appends the string "battery" for the battery level value. Example - `5% battery`

WDYT?

AnirudhBhat commented:
Thanks for working on this @samiuelson The functionality works as expected 🎉 

I've left a few minor comments. Again, Great job on this!

samiuelson committed:
Test battery level not added to event if null

samiuelson committed:
Test battery level not updated if didn't change

samiuelson committed:
Track card reader battery level in percents

Extracts util functions into separate file and reuses them in `CardReaderDetailViewModel` and `CardReaderTracker`.

samiuelson commented:
> Thanks for working on this @samiuelson The functionality works as expected 🎉
> 
> I've left a few minor comments. Again, Great job on this!

Thanks for the review @AnirudhBhat! I've added suggested changes 👍🏼

AnirudhBhat commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderUtils.kt:
```diff
@@ -0,0 +1,27 @@
+import kotlin.math.roundToInt
+
+private const val PERCENT_100 = 100
+private fun Float.toPercent(): Int = (this * PERCENT_100).roundToInt()
```
👍

AnirudhBhat commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackerTest.kt:
```diff
@@ -101,6 +103,17 @@ class CardReaderTrackerTest : BaseUnitTest() {
         }
 
+    @Test
+    fun `given battery level is null, when event is tracked, then it shouldn't contain battery level property`() =
```
❓ We could write another test that verifies that the battery level property is included when the battery level is not null. WDYT?

```
@Test
    fun `given battery level is not null, when event is tracked, then it should contain battery level property`() =
        testBlocking {
            val cardReaderTrackingInfo: TrackingInfo = mock {
                on { cardReaderBatteryLevel } doReturn 0.5F
            }
            whenever(cardReaderTrackingInfoProvider.trackingInfo).thenReturn(cardReaderTrackingInfo)
            val props = mutableMapOf<String, Any>()

            cardReaderTracker.track(CARD_PRESENT_COLLECT_PAYMENT_SUCCESS, props)

            assertTrue(props.containsKey("card_reader_battery_level"))
        }
```

AnirudhBhat reviewed (approved):
Thanks for the changes @samiuelson. I've left one minor comment. LGTM!

samiuelson committed:
Test that battery level is tracked with events
